## Inferential Statistics

While using RFE we discussed that it uses `p-value` to select the most relevant features. Now we will learn more about it. To understand `p-values`, we first need to learn about basics in Inferential Statistics and Hypothesis Testing.

- **Descriptive statistic**: It is the branch of classical statistics that describes / summarizes the data through some statistics which include mean, median, mode, frequency, standard deviation, and variance. In descriptive statistics, there is no uncertainty as we calculate these statistics on the measured values.

- **Inferential statistics**: It is the branch of classical statistics that helps us to draw inferences about the population, from the given sample data. The basic goal of inferential stats is to draw conclusions from a sample (smaller number of subjects/samples) and generalize them on a larger population (where the complete data is unavailable and we have only a sample of data)

### Hypothesis Testing
It is one of the tools that is used to make inferences on the population (which is called our hypothesis), from the sample data.

Elements of hypothesis Tests:

- **Null hypothesis**: Denoted with H0, a null hypothesis is an assumption that the population average is identical to a specific value. The typical notation is μ = μ0, where μ refers to the population mean and μ0 refers to the hypothesized value.
- **Alternate hypothesis**: An alternative hypothesis is the opposite of the null hypothesis. We compare this hypothesis with the null hypothesis to decide whether or not we reject the null hypothesis. We denote the alternative hypothesis with H1 or Ha.
- **Level of significance**: we use level of significance to indicate whether we are confident enough to reject the null hypothesis.
- **Test Statistic**: Once we determine the type of hypothesis test and that our assumptions have been met, we use our data to decide whether to reject or not reject the null hypothesis.(z-test or t-test)
- **p-value**: is a measure used help us to reject or accept the null hypothesis


### Problem Statement
Boys of a certain age are known to have a mean weight of μ = 85 pounds. A complaint is made that the boys living in a municipal children's home are underfed. As one bit of evidence, n = 25 boys (of the same age) are weighed and found to have a mean weight of 80.94 pounds. It is known that the population standard deviation σ is 11.6 pounds (the unrealistic part of this example!). Based on the available data, what should be concluded concerning the complaint?

**What are we trying to establish here?**
It is assumed that the population mean weight is 85lbs, but we do not have the complete data from the population. Otherwise we would have calculated the actual mean directly. However we only have sample data from 25 subjects. So based on this sample data we will try to prove or disprove our assumption, using statistical test.


**Steps**
- Step1: Define the null hypothesis - This is our assumption about the population. It is defined by H0 and in this case `H0: μ = 85`;
- Step 2: Define the alternative hypothesis - This means, what if our assumption is not true. It is defined by Ha and in this case `Ha: μ != 85`.
    Since in this case, our null hypothesis would be false on either side of `Ha`, hence such a test is called a **Two tailed Test**, ie., if `Ha: μ > 85` and if `Ha: μ < 85`, in both cases, `H0` is false.
- Step 3: Decide a **test statistics** based on the information available. Assuming data is normally distributed and number of observations are less and **population variance is known (since population standard deviation is provided), we will use a `z-test`**. This test is based on a "z-distribution" which is the same as a standard normal distribution. If the population variance was not known, we would have used a t-test. T test is based on `students t distribution` which is very similar to a standard normal distribution except that it is much flatter: [link to the image - T distribution vs Standard Normal Distribution](https://education-team-2020.s3-eu-west-1.amazonaws.com/data-analytics/7.03/7.03-t_distribution.png)
- Step 4: Level of significance: This defines the rejection region/critical region. It is defined by greek letter 'alpha'. Usually it is 0.05.

Explain the significance of alpha: [link to the image - Two Tailed Test Rejection Region](https://education-team-2020.s3-eu-west-1.amazonaws.com/data-analytics/7.03/7.03-two_tailed_test.png)

  - Step 5: Calculate the test statistic based on the given information

    - If the test statistic falls in the critical region, then we reject the Null Hypothesis
    - If the test statistic falls in the region between the critical region, then we fail to reject the Null Hypothesis, which means we accept the Null Hypothesis ie our assumption about the population to be true.

<img src=https://cdn.analyticsvidhya.com/wp-content/uploads/2020/03/Screenshot-from-2020-03-04-15-29-37.png width="500">

In [1]:
from statistics import math

sample_mean = 80.94
pop_mean = 85
pop_std = 11.6
n = 25
statistic = (sample_mean - pop_mean)/(pop_std/math.sqrt(n))
print("Statistic is: ", statistic)

Statistic is:  -1.750000000000001


- If the test statistic falls in the critical region, then we reject the Null Hypothesis
- If the test statistic falls in the region between the critical region, then we fail to reject the Null Hypothesis, which means we accept the Null Hypothesis ie our assumption about the population to be true

- This critical region is defined by a critical value which is symmetrical on either side of the Y axis.
- We use the significance level value, alpha, to check the critical value using a t-table. You can find the t table here on this link [https://www.statisticshowto.com/tables/t-distribution-table/#two]. Since this is a two tailed test, we will use `alpha/2` ie 0.025 and see that the critical value is 2.064
  **Here is a simple [video](https://www.youtube.com/watch?v=6CwtZjvy7N4) that the students can refer, to understand how to find the `z-critical` value for a confidence interval. In this case alpha is 0.05 (confidence level is 95%).**
- Now we can compare the calculated statistic and the critical value to see if the null hypothesis is rejected or not. In this case the statistic is calculated to be -1.75 and the critical value is 2.064 (or -2.064), hence we are not in the rejection region. Hence we fail to reject the null hypothesis. </br>
[Find the statistic here](https://www.statisticshowto.com/tables/t-distribution-table/#two)

### p-value

- Using P values is another way of rejecting or failing to reject the Null Hypothesis. We can use it to verify the previous result, ie whether we reject the Null Hypothesis or not. Then we will discuss the meaning of `p-values`

  - Reminder: `t-statistic = (sample_mean - nullhypothesis_mean)/ (sample_stddev / sqrt(n) )`
  - whereas: `nullhypothesis_mean` is the population mean
  - Find the p-value from [the table](https://www.ztable.net/) and verify the result.

  **Method:** In this case the statistic was calculated to be -1.75. Now look in the z table and find the corresponding p-value. Here it would be 0.04
  This is greater than 0.025 (since it is a two sided test, alpha is split symetrically on either side of the y axis), **hence he fail to reject the null hypothesis** and accept the alternative hypothesis.

- What is a p-value: Technically p-value is a probability value of rejecting the null hypothesis or failing to reject the null hypothesis. It is used as a direct measure of statistical significance, which we would see later.
- p-value = what is the probability, measuring P(H1 | H0 is true), or in our case P(finding a sample mean of 80.94 of weight among boys | the boys weight results out of pure random selection - and no underfeeding whatsoever)

### Two sample t-test

- Here instead of drawing an inference on one population through its sample data, we try to compare the means of two populations from which two different sample sets are drawn. It helps us to test whether the population means of the two groups are equal or not.

This time we will use a T test because, as we would see later, population variances are unknown here. The way the hypothesis/experiment is set up is the same as the process used in the previous examples with Z test.

**It can used to analyze the results from A/B test as well**.

- Some of the assumptions on the data before we can use this test are:
  - Populations from which the samples are drawn is assumed to be **normally distributed**
  - Sample data has **continuous values** and **not discrete**
  - Random samples must be drawn from the populations
  - Data in the two samples must be independent of each other

#### Problem statement

- A psychologist was interested in exploring whether or not male and female college students have different driving behaviors. There were a number of ways that she could quantify driving behaviors. She opted to focus on the fastest speed ever driven by an individual. Therefore, the particular statistical question she framed was as follows:
  - Is the mean fastest speed driven by male college students different than the mean fastest speed driven by female college students?
  - She conducted a survey of a random n = 34 male college students and a random m = 29 female college students. Here is a descriptive summary of the results of her survey:

```shell
# Males
samples = 34
Sample mean = 105.5
Sample standard deviation: 20.1

# Females
samples = 29
Sample mean = 90.0
Sample standard deviation:12.2
```

- We start with calculating the pooled sample variance: [link to the image - Pooled Sample Variance](https://education-team-2020.s3-eu-west-1.amazonaws.com/data-analytics/7.04/7.04-pooled_sample_variance.png)
- Next we calculate the statistic using the below formula: [link to the image - Pooled T Test - Statistic](https://education-team-2020.s3-eu-west-1.amazonaws.com/data-analytics/7.04/7.04-pooled_t_test.png)


In [2]:
sample_mean1 = 105.5
sample_std1 = 20.1
n1 = 34

sample_mean2 = 90.9
sample_std2 = 12.2
n2 = 29

T Statistic is:  3.4101131776909535


In [5]:
from scipy.stats import ttest_ind, norm

# create the samples
males = norm.rvs(loc=sample_mean1, scale=sample_std1, size=n1)
females = norm.rvs(loc=sample_mean2, scale=sample_std2, size=n2)

In [6]:
males

array([ 68.4248047 , 104.75748979,  89.23263148, 106.07735938,
        81.7746044 ,  99.11714014, 108.08416928,  97.67086084,
       118.53943973, 126.64763954, 120.08668291,  87.52912184,
        86.21676983, 100.13232159,  92.55652321, 142.91661636,
       107.96348335,  73.42011799,  92.97660102,  63.44947364,
       102.14480484, 116.81137863, 112.90006953,  97.59391166,
       100.13881478, 134.03220532, 148.20056766, 103.15143838,
       108.14430919,  67.63225672, 118.11996858, 127.00122468,
       108.24720908, 134.21293753])

In [7]:
females

array([ 85.70064177,  92.48742651, 124.35531002,  73.55316896,
       105.85692447, 104.09608077,  74.07575375,  89.64300039,
        86.09690305,  83.38452122, 115.50668514,  99.58866847,
        77.64217242,  82.92683857,  78.85130096,  84.18712929,
       103.97227856,  76.22357577,  83.82005813,  74.6404814 ,
        81.79834993,  98.94841192,  94.05419213,  72.38486846,
        90.51102871,  76.51719217,  89.48760023,  87.53804844,
        94.82128141])

In [8]:
ttest_ind(males, females)

Ttest_indResult(statistic=3.4304490603466724, pvalue=0.0010865240745535569)

We reject the null hypothesis. </br>
We can say with 95% of confidence that we have enough evidence to discard the null hypothesis.</br>
The average speed between male and females are not the same.</br>
Comparing the means, we can say males drive faster.